vchan/init.c

```diff
@@ -29,30 +29,29 @@
 #include "libvchan_private.h"
 
 libvchan_t *libvchan_server_init(int domain, int port, size_t read_min, size_t write_min) {
-    char xs_path[255];
     libvchan_t *ctrl;
 
-    ctrl = malloc(sizeof(*ctrl));
+    ctrl = calloc(sizeof(*ctrl), 1);
     if (!ctrl)
         return NULL;
 
-    snprintf(xs_path, sizeof(xs_path), "data/vchan/%d/%d", domain, port);
-    ctrl->xenvchan = libxenvchan_server_init(NULL, domain, xs_path, read_min, write_min);
-    if (!ctrl->xenvchan) {
-        free(ctrl);
-        return NULL;
-    }
-    ctrl->xs_path = strdup(xs_path);
+    if (asprintf(&ctrl->xs_path, "data/vchan/%d/%d", domain, port) < 0)
+        goto err_asprintf;
+    ctrl->xenvchan = libxenvchan_server_init(NULL, domain, ctrl->xs_path, read_min, write_min);
+    if (!ctrl->xenvchan)
+        goto err_server_init;
     ctrl->xenvchan->blocking = 1;
     ctrl->remote_domain = domain;
-    ctrl->xc_handle = xc_interface_open(NULL, NULL, 0);
-    if (!ctrl->xc_handle) {
-        /* error already logged by xc_interface_open */
-        libxenvchan_close(ctrl->xenvchan);
-        free(ctrl);
-        return NULL;
-    }
-    return ctrl;
+    if ((ctrl->xc_handle = xc_interface_open(NULL, NULL, 0)))
+        return ctrl; /* Everything good */
```

Comment:
I'd prefer conditional handling an error, and final successful `return` be outside of `if`. It means one more goto label, but IMO is easier to read.